library/core/src/num/mod.rs

```diff
@@ -846,37 +870,30 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+
+    if radix <= 16 && src.len() <= mem::size_of::<T>() * 2 - (is_signed_ty && is_positive) as usize
+    {
+        // SAFETY: Consider the highest radix of 16:
+        // `u8::MAX` is `ff` which is 2 characters, `u16::MAX` out is `ffff` which is 4 characters
+        // We can be sure that any src len of 2 would fit in a u8 so we don't need
+        // to check for overflow.
+        unsafe {
+            let additive_op = if is_positive { T::unchecked_add } else { T::unchecked_sub };
+
+            for &c in digits {
+                result = result.unchecked_mul(radix);
+                let x = (c as char).to_digit(radix).ok_or_else(|| PIE { kind: InvalidDigit })?;
+                result = additive_op(&result, x);
+            }
         }
     } else {
-        // The number is negative
+        let additive_op = if is_positive { T::checked_add } else { T::checked_sub };
+
         for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: NegOverflow }),
-            };
-            result = match result.checked_sub(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: NegOverflow }),
-            };
+            let mul = result.checked_mul(radix);
+            let x = (c as char).to_digit(radix).ok_or_else(|| PIE { kind: InvalidDigit })?;
+            result = mul.ok_or_else(|| PIE { kind: PosOverflow })?;
+            result = additive_op(&result, x).ok_or_else(|| PIE { kind: PosOverflow })?;
```

Comment:
In the checked case the difference isn't just `add` vs `sub` but also `PosOverflow` vs `NegOverflow`.

Reply:
D'oh. Good catch.

Reply:
Hmm, if you can spot a nicer way to do it I'm all ears. It seems it's not a copy type - maybe IntErrorKind should be a Copy type?

Reply:
See my comment on the current code.

Reply:
excellent - that tidies it up nicely.